### PR TITLE
[DependencyInjection] Prevent service:method factory notation in PHP config

### PR DESCRIPTION
**Pull request by vudaltsov**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #24704 
| License       | MIT
| Doc PR        | 

Started working on fixing #24704.

@nicolas-grekas, am I on the right way? If yes I will look at the tests and try to add this case.

### CODE REVIEW

**Comment by vudaltsov** [comment]
Should I move the string splitting logic to the `AbstractConfigurator::processValue`?

**Comment by ro0NL** [comment]
See `YamlFileLoader::parseCallable`; we need to support `configurator` as well i guess. Also account for `::` before splitting by `:` (see `setFactory`).

Im not sure why we dont do this directly in `Definition::setFactory` & co; right now XML is missing this feature as well. (e.g. `function="service:method"`).

Otherwise perhaps go with `static::processValue(static::processCallable())` or so.

**Comment by nicolas-grekas** [comment]
I know I suggested this one, but I've mixed feelings about it :)
What about throwing instead? Providing a single way to reference services would make teaching easier.
The message should suggest to use `[ref('foo'), 'bar']` when `foo:bar` is provided.
WDYT?

**Comment by ogizanagi** [comment]
I'd also prefer the exception suggested by @nicolas-grekas instead of this.

**Comment by vudaltsov** [comment]
You know, I would also agree :)
When trying to implement it I realized that it looks unnatural.
`[ref(), '']` definitely makes much more sense.

**Comment by nicolas-grekas** [comment]
@vudaltsov let's throw here then instead?

**Comment by vudaltsov** [comment]
@nicolas-grekas, done

File: src/Symfony/Component/DependencyInjection/Loader/Configurator/Traits/FactoryTrait.php [hunk 1]
```diff
@@ -22,6 +24,10 @@
      */
     final public function factory($factory)
     {
+        if (is_string($factory) && 1 === substr_count($factory, ':')) {
+            throw new InvalidArgumentException('The short factory notation is not supported by the PHP-based DI configuration.');
```
Comment (ogizanagi) [inline src/Symfony/Component/DependencyInjection/Loader/Configurator/Traits/FactoryTrait.php]: In addition, it should also suggest to use `[ref('service'), 'method']` instead.

Comment (vudaltsov) [reply src/Symfony/Component/DependencyInjection/Loader/Configurator/Traits/FactoryTrait.php]: So the final exception will look like `The service:method factory notation is not supported by the PHP-based DI configuration. Use [ref('service'), 'method'] instead.`

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Loader/Configurator/Traits/FactoryTrait.php]: LGTM, with double quotes around, and resolved/actual id+method, eg:
```
Invalid factory "%s:%s": the `service:method` notation is not available when using PHP-based DI configuration. Use "[ref('%s'), '%s']" instead.
```

Comment (ro0NL) [reply src/Symfony/Component/DependencyInjection/Loader/Configurator/Traits/FactoryTrait.php]: Static is still allowed using `factory('class::method')`? Or should that notation also be explicit; thus `factory('class', 'method')`.

Comment (vudaltsov) [reply src/Symfony/Component/DependencyInjection/Loader/Configurator/Traits/FactoryTrait.php]: @nicolas-grekas, done

**Review by nicolas-grekas — APPROVED** [review]
(could you please squash, and update the commit message meanwhile please)

**Comment by vudaltsov** [comment]
@nicolas-grekas, squashed, reworded and rebased

**Comment by nicolas-grekas** [comment]
Thank you @vudaltsov.

**Pull request merged**